### PR TITLE
Add tests and correct TouchTask behavior

### LINKED ISSUE #1452: [TouchTask] allows invalid datetime values to be specified
The two examples below are not flagged by TouchTask as invalid even though it cannot use them.

`<touch file="somefile" datetime="Not A DateTime" />`
`<touch file="somefile" datetime="20201127:1123" />`

### PR DESCRIPTION
**Pull request by BlairCooper**
Fixes #1452, #1453, #1454

- Check return value of strtotime() - #1452
- Change millis member of TouchTask to seconds to reflect what it
actually contains. Retain millis attribute for continuity with the ANT
Touch task - #1453
- Add tests for validating timestamps and error conditions - #1454

### CODE REVIEW

File: test/etc/tasks/system/TouchTaskTest.xml [hunk 1]
```diff
@@ -52,11 +54,59 @@
...
     <target name="testMappedFilelist">
-        <touch datetime="${DSTAMP}:${TSTAMP}">
+        <touch datetime="${formattedTimestamp}">
             <filelist dir="." files="${tmp.dir}/idonotexist" />
             <mapper type="merge" to="${tmp.dir}/touchtest" />
         </touch>
     </target>
 
+    <target name="testMillis">
+        <touch file="${tmp.dir}/millis-file" millis="946684799864" /> <!-- Friday, December 31, 1999 11:59:59.864 PM -->
```
Comment (siad007) [inline test/etc/tasks/system/TouchTaskTest.xml:64]: Please test against milliseconds not microseconds.

File: test/classes/phing/tasks/system/TouchTaskTest.php [hunk 2]
```diff
@@ -1,4 +1,6 @@
 <?php
+use PHPUnit\Framework\Constraint\IsEqual;
```
Comment (siad007) [inline test/classes/phing/tasks/system/TouchTaskTest.php:2]: Please move after file comment!

File: test/classes/phing/tasks/system/TouchTaskTest.php [hunk 3]
```diff
@@ -106,4 +108,127 @@ public function testMappedFilelist()
...
+     */
+    public function testMillis()
+    {
+        // Don't run the test on 32-bit systems
+        if (PHP_INT_SIZE > 4) {
+            $this->executeTarget(__FUNCTION__);
+            $testFile = $this->getProject()->getProperty('tmp.dir') . '/millis-file';
+            $this->assertFileExists($testFile);
+
+            $this->assertEquals('December 31 1999 23:59:59', date("F d Y H:i:s", filemtime($testFile)));
```
Comment (siad007) [inline test/classes/phing/tasks/system/TouchTaskTest.php:123]: Do you mean something like this?
```php
$this->assertEquals('December 31 1999 23:59:59.000000', date("F d Y H:i:s.u", filemtime($testFile)));
```

**Review by siad007 — CHANGES_REQUESTED** [review]

File: test/etc/tasks/system/TouchTaskTest.xml (hunk 1 above)
Comment (BlairCooper) [reply test/etc/tasks/system/TouchTaskTest.xml:64]: It is milliseconds:
image

File: test/classes/phing/tasks/system/TouchTaskTest.php [hunk 4]
```diff
@@ -106,4 +108,127 @@ public function testMappedFilelist()
...
+        $testFile = $this->getProject()->getProperty('tmp.dir') . '/default-now-file';
+        $this->assertFileExists($testFile);
+
+        /*
+         * Assert that the timestamp is within 1 second of the time the test
+         * started. Ideally it's exactly the same but we'll allow for minimal
+         * drift to account for a lag between when we noted the time and when
+         * the file was touched.
+         */
+        $this->assertThat(filemtime($testFile), new IsEqual($nowTime, 1), 'File timestamp not within 1 second of now');
```
[the remark this answers is not in the document]
Comment (BlairCooper) [reply test/classes/phing/tasks/system/TouchTaskTest.php:232]: Done

File: test/classes/phing/tasks/system/TouchTaskTest.php (hunk 2 above)
Comment (BlairCooper) [reply test/classes/phing/tasks/system/TouchTaskTest.php:2]: Done

File: test/classes/phing/tasks/system/TouchTaskTest.php (hunk 3 above)
Comment (BlairCooper) [reply test/classes/phing/tasks/system/TouchTaskTest.php:123]: Seems unnecessary as filemtime() would always return the milliseconds as .000000.

**Review by siad007 — CHANGES_REQUESTED** [review]
@BlairCooper please check https://github.com/phingofficial/phing/pull/1456/checks?check_run_id=1513668571#step:5:633
```
[phpunit] Testsuite: TimerTest
  [phpunit] Tests run: 1, Risky: 0, Warnings: 0, Failures: 1, Errors: 0, Incomplete: 0, Skipped: 0, Time elapsed: 0.01496 s
  [phpunit] testTimer FAILED
  [phpunit] Failed asserting that '0.00194' matches expected 0.0.
```

**Review by siad007 — APPROVED** [review]

**Pull request merged**